lib/api/core/vault.js

```diff
@@ -0,0 +1,211 @@
+   *
+   * Rejects with an error if:
+   *  - a key is found but no file
+   *  - a file is found but no key
+   *
+   * @param {string|null} secretsFile - file containing the encrypted secrets
+   * @returns {Promise}
+   */
+  init (secretsFile = null) {
+    this.encryptedSecretsFile = secretsFile || process.env.KUZZLE_SECRETS_fILE|| defaultSecretsFile;
+
+    const fileExists = fs.existsSync(this.encryptedSecretsFile);
+
+    if (!this.vaultKey && !fileExists) {
+      return Promise.resolve();
+    }
+
+    if (this.vaultKey && !fileExists) {
+      return Promise.reject(
+        new Error('A vault key is present and no secrets file can be found. Aborting.')
+      );
+    }
+
+    if (!this.vaultKey && fileExists) {
+      return Promise.reject(
+        new Error('A secrets file is present and no vault key can be found. Aborting.')
+      );
+    }
+
+    return this._readJsonFile(this.encryptedSecretsFile)
+      .then(encryptedSecrets => this.decryptObject(encryptedSecrets))
+      .then(secrets => {
+        this.secrets = secrets;
+
+        return null;
+      })
+      .catch(error => {
+        throw new Error(`Can not decrypt secrets: ${error.message}`);
+      });
+  }
+
+  /**
+   * Iterate recursively through object values and try to
+   * decrypt strings only.
+   *
+   * @param {object} encryptedSecrets - object containing the encrypted secrets
+   */
+  decryptObject (encryptedSecrets) {
+    if (!this.vaultKey) {
+      throw new Error('Can not find vault key. Aborting.');
+    }
+
+    const secrets = {};
+
+    for (const key of Object.keys(encryptedSecrets)) {
+      const value = encryptedSecrets[key];
+
+      if (_.isPlainObject(value)) {
+        secrets[key] = this.decryptObject(value);
+      } else if (_.isString(value)) {
+        secrets[key] = this._decryptData(value);
+      }
+    }
+
+    return secrets;
+  }
+
+  /**
+   * Iterate recursively through object values and try to
+   * encrypt strings only.
+   *
+   * @param {object} secrets - object containing the secrets
+   */
+  encryptObject (secrets) {
+    if (!this.vaultKey) {
+      throw new Error('Can not find vault key. Aborting.');
+    }
+
+    const encryptedSecrets = {};
+
+    for (const key of Object.keys(secrets)) {
+      const value = secrets[key];
+
+      if (_.isPlainObject(value)) {
+        encryptedSecrets[key] = this.encryptObject(value);
+      } else if (_.isString(value)) {
+        encryptedSecrets[key] = this._encryptData(value);
+      }
+    }
+
+    return encryptedSecrets;
+  }
+
+  /**
+   * Encrypt data with AES CBC using the secret key and an initialization vector
+   * It's not safe to re-use an IV , so we generate a new IV each time we encrypt
+   * something and we store it next to the encrypted data.
+   * See https://www.wikiwand.com/en/Block_cipher_mode_of_operation#/Initialization_vector_(IV)
+   */
+  _encryptData (data) {
+    const
+      IV = crypto.randomBytes(8).toString('hex'),
```

Comment:
(nitpicking) names starting with an uppercase should be reserved to classes

```suggestion
      iv = crypto.randomBytes(8).toString('hex'),
```

(same thing in the _decryptData function below)